### DISCUSSION TITLE
Rustup settings on windows for esp32 target only?

### DISCUSSION THREAD

**Discussion by JasonKleban** [esp-rs/espup#313] [Q&A] [answered]
I use a windows laptop for casual esp32 development. Begrudgingly, I use micropython and Thonny IDE only because I haven't figured out a good rust setup, or even a reliable mpy on vscode flow.  I'm ready to try rust again this year.  I'm mostly new to embedded and I'm sure it's apparent that I'm fuzzy on the details.

espup seems to require rustup installed to install.  rustup installation wants gcc or msvc tooling and asks for a host triple, which is the target architecture of code builds, right?  So esp32 isn't a top-tier yet, can I still exclusively pick that? Or we have to pick a top tier and add esp32 only as secondary?

I don't want to install toolchain, etc tools that aren't necessary and which might confuse me or my system later.  I don't want to complicate my setup with WSL or docker containers or I fear I'll spend the time futzing with that instead of the esp32 - when it doesn't seem that it would be necessary.  I just haven't found the right and updated guide.

Do I really need gcc/msvc or does the xtensa and esp32 have enough tools to keep rust updated, build for esp32, write to the chip and debug?  What host-triple is the best pick for only targeting esp32 series chips?  What default toolchain is the best choice for this?

Otherwise I'm expecting to use vscode with rs-probe and serial-monitor plugins and the rust language service and following recent guides on, say https://apollolabsblog.hashnode.dev but a lot of these resources (I think?) assume a linux workstation.

Thanks!

P.S. since everyone seems to be on laptops these days, and most developers aren't on apple hardware, and non-apple laptop hardware has notoriously poor linux driver coverage, why are rust-on-windows developers seemingly so uncommon?

**Comment by SergioGasquez** [comment, marked as answer]
Hi! 

> I use a windows laptop for casual esp32 development. Begrudgingly, I use micropython and Thonny IDE only because I haven't figured out a good rust setup, or even a reliable mpy on vscode flow. I'm ready to try rust again this year. I'm mostly new to embedded and I'm sure it's apparent that I'm fuzzy on the details.

I would suggest taking a look at https://esp-rs.github.io/book/tooling/visual-studio-code.html

> espup seems to require rustup installed to install. rustup installation wants gcc or msvc tooling and asks for a host triple, which is the target architecture of code builds, right? So esp32 isn't a top-tier yet, can I still exclusively pick that? Or we have to pick a top tier and add esp32 only as secondary?

Basically, `espup` requires `rustup` and rustup requires an ABI, that is GCC or MSVC, see https://esp-rs.github.io/book/installation/rust.html. You can't choose esp32 for host triple, you need to install those tools for your host machine (esp32 is the target machine), in your case windows x64.

> I don't want to install toolchain, etc tools that aren't necessary and which might confuse me or my system later. I don't want to complicate my setup with WSL or docker containers or I fear I'll spend the time futzing with that instead of the esp32 - when it doesn't seem that it would be necessary. I just haven't found the right and updated guide.

There are a few options:
- [You install the environment in your machine](https://esp-rs.github.io/book/installation/index.html)
- You use online [Wokwi simulator](https://wokwi.com/rust)
- You use [containers](https://esp-rs.github.io/book/installation/using-containers.html) or [devcontainers](https://esp-rs.github.io/book/writing-your-own-application/generate-project/index.html#using-dev-containers-in-the-templates)
  - Unfortunately, this also requires some WSL/docker setup on Windows
 
> P.S. since everyone seems to be on laptops these days, and most developers aren't on apple hardware, and non-apple laptop hardware has notoriously poor linux driver coverage, why are rust-on-windows developers seemingly so uncommon?

I think using Rust on Windows is perfectly fine, even some of our teams members use it for work!

**Reply by JasonKleban** [reply]
Thanks!

> You can't choose esp32 for host triple, you need to install those tools for your host machine (esp32 is the target machine), in your case windows x64.

I was partly basing my truth on [this accepted answer](https://stackoverflow.com/a/49368388/162273)  "Host triples identify the architecture and OS of the system that will ultimately run your executable."  Is that wrong?  Or am I missing a nuance?

I'm not _entirely_ against the docker devcontainer if it really is the smoothest.  I've seen mention of it, but I haven't gotten deep enough into it to find a step by step guide.  I'll explore your links.  Thanks!

**Reply by SergioGasquez** [reply]
> I was partly basing my truth on [this accepted answer](https://stackoverflow.com/a/49368388/162273) "Host triples identify the architecture and OS of the system that will ultimately run your executable." Is that wrong? Or am I missing a nuance?

Yes, that's not exactly true. In your case, you need to use Windows x64 toolchains to crosscompile to ESP32  

>  I'm not entirely against the docker devcontainer if it really is the smoothest. I've seen mention of it, but I haven't gotten deep enough into it to find a step by step guide. I'll explore your links. Thanks!

IMHO, the smoothest is to install everything on your machine, but if you want to have a minimal environment, then devcontainers might be a safe option
